### PR TITLE
Improve memoization of LHN by handling the isFocused prop better

### PR DESCRIPTION
**Pull request by kacper-mikolajczak**
## Intro

Transition between chats (reports), seems to be quite laggy. In this document, we try to investigate the issue and come up with possible solution. 

## Steps to reproduce

1. Load two chats
2. Navigate from one to another

## Recording

https://github.com/Expensify/App/assets/62747088/85bb45e8-e394-4f61-bc8a-5fe57361ef77

## Investigations

It looks like there is a big overhead in form of rendering LHN’s list of `OptionRowLHNData` components.
BeforeFixes

In the profiling trace we see that every item in the list is re-rendered. Each re-render takes up to **5ms** to complete, which with current list size can significantly throttle rendering.

The re-rendering is caused by change in `currentReportID` state, which is provided by `withCurrentReportID` HOC. `currentReportID` prop is used in  `OptionRowLHNData` in one place, to find out wether component should be focused or not:

```tsx
const isFocused = !shouldDisableFocusOptions && currentReportID === reportID;
```

The `isFocused` is then directly passed down into child.

The change in `currentReportID` breaks memoization, causing heavy re-render of each item.

### Wrapper component

As `currentReportID` is only used **********once********** and `isFocused` is a boolean flag, we can extract this logic into separate **************wrapper************** component and refactor `OptionRowLHNData` to receive primitive `isFocused` prop, which will not break memoization. 

AfterFixes copy

By applying these changes, we were able to shrink down the rendering time of List component to **~100ms**, which will be even more significant as number of list items grows. The rest of rendering overhead seems to be related to different components, `Report` itself etc.

The rendering time should be now only merely related to list size, as it only triggers one heavy re-render.

Wrapper component would look more or less like this:

```tsx
function OptionRowLHNDataWrapper({currentReportID, shouldDisableFocusOptions, ...props}) {
    const isFocused = !shouldDisableFocusOptions && currentReportID === props.reportID;
    return (
        <OptionRowLHNDataComposed
            {...props}
            isFocused={isFocused}
        />
    );
}
```

This solution does not add anything beyond a simple wrapper so it is quite easy to do and should not be error-prone.

### Use `withCurrentReportID` in list component (Alternative)

AfterFixesDetailedAlternatives copy

Alternatively, we can use `withCurrentReportID` once at the level of list component instead of in each children. Then we will directly pass `isFocused` prop into children, so it won’t break memoization of list items.

The solution is more elegant, as it uses `currentReportID` provider once and does not expose such internals into children components. On the flipside, we need to re-render list itself, which might add some overhead.

In this scenario, it takes about **********~120ms********** to render the list, which is about **********15ms********** slower than previous solution.

## Conclusions

While we managed to reduce rendering times, we still need to think on which solution should be implemented. Furthermore, it might be a good idea to investigate rendering of list items, as it is quite simple component but still takes ************~5ms************ to render.

### Fixed Issues

$ https://github.com/Expensify/App/issues/28631
PROPOSAL: https://github.com/Expensify/App/issues/28631 (it contains a proposal)

### Tests

- [x] Verify that no errors/warnings appear in the JS console.

1. Open app
2. See if can easily switch between multiple chats
3. See if focus (opacity) changes according to selected chats

### Offline tests

N/A
### QA Steps

1. Open app
2. See if can easily switch between multiple chats
3. See if focus (opacity) changes according to selected chats

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/62747088/54d241b3-210a-4f0a-ab46-69b4cb0775a8

</details>

<details>
<summary>Mobile Web - Chrome</summary>

N/A

</details>

<details>
<summary>Mobile Web - Safari</summary>

N/A

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/62747088/e4320146-5aa4-410c-b4d7-cc14a749f96f

</details>

<details>
<summary>iOS</summary>

Mobile recording for reference

https://github.com/Expensify/App/assets/62747088/9f24267c-0a2a-421e-9cd8-a7dad7134b31

</details>

<details>
<summary>Android</summary>

N/A

</details>

### CODE REVIEW

**Comment by kacper-mikolajczak** [comment]
Hi @muttmuure, here is a draft of partial improvements that I've done recently. I wanted to share it on Slack but it is quite hard to put such document into Slack message :D Let me know what you think about results and overall shape of such document. In the meantime I will work on further improvements

**Comment by mountiny** [comment]
@kacper-mikolajczak Thanks! I think this looks great!

I agree that the second solution leads to more readable code, adding that one wrapper feels a bit odd to me.

@kacper-mikolajczak could you summarize these two options and post it to #newdot-performance room? I think it would be great to get opinions of other agency experts and engineers on this and then we can implement this. Thanks!

**Comment by kacper-mikolajczak** [comment]
@mountiny  Yeah definitely! Thanks for the feedback 👍

File: src/components/LHNOptionsList/OptionRowLHNData.js [hunk 1]
```diff
@@ -20,8 +19,8 @@ import CONST from '../../CONST';
 import reportActionPropTypes from '../../pages/home/report/reportActionPropTypes';
 
 const propTypes = {
-    /** If true will disable ever setting the OptionRowLHN to focused */
-    shouldDisableFocusOptions: PropTypes.bool,
+    /** Wether row should be focused */
```
Comment (hurali97) [inline src/components/LHNOptionsList/OptionRowLHNData.js:22]:
```suggestion
    /** Whether row should be focused */
```

File: src/components/LHNOptionsList/OptionRowLHNData.js [hunk 2]
```diff
@@ -91,7 +86,6 @@ function OptionRowLHNData({
     const reportID = propsToForward.reportID;
     // We only want to pass a boolean to the memoized component,
     // instead of a changing number (so we prevent unnecessary re-renders).
-    const isFocused = !shouldDisableFocusOptions && currentReportID === reportID;
```
Comment (hurali97) [inline src/components/LHNOptionsList/OptionRowLHNData.js:94]: We can move this comment to `OptionRowLHNDataWithFocus`, since we moved the `isFocused` there as well.

File: src/components/LHNOptionsList/OptionRowLHNDataWithFocus.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    ...withCurrentReportIDPropTypes,
+    shouldDisableFocusOptions: PropTypes.bool,
+};
+
+const defaultProps = {
+    ...withCurrentReportIDDefaultProps,
+    shouldDisableFocusOptions: false,
+};
+
+function OptionRowLHNDataWithFocus({currentReportID, shouldDisableFocusOptions, ...props}) {
```
Comment (hurali97) [inline src/components/LHNOptionsList/OptionRowLHNDataWithFocus.js:16]: Can we add some comments stating what this `OptionRowLHNDataWithFocus ` does? It would be cool in future to have an understanding of what it does because not everyone will be coming back to the PR to understand.

**Review by hurali97 — CHANGES_REQUESTED** [review]
LGTM ! Just a couple comments.

File: src/components/LHNOptionsList/OptionRowLHNData.js (hunk 1 above)
Comment (kacper-mikolajczak) [reply src/components/LHNOptionsList/OptionRowLHNData.js:22]: Done ✅

File: src/components/LHNOptionsList/OptionRowLHNDataWithFocus.js (hunk 3 above)
Comment (kacper-mikolajczak) [reply src/components/LHNOptionsList/OptionRowLHNDataWithFocus.js:16]: Done ✅

**Comment by kacper-mikolajczak** [comment]
BTW, on iPhone 13 simulator the difference is huge

**Comment by mountiny** [comment]
Damn, that looks great!

**Comment by kacper-mikolajczak** [comment]
@mountiny Yeah that does look promising! Wonder what would be a perceived perf gain, as the list is out of the view when the render happens (it is covered by report screen). But hey, at least we reduced the resources consumption a lil bit :D

**Comment by kacper-mikolajczak** [comment]
Sure @mountiny, testing it right now!

**Comment by kacper-mikolajczak** [comment]
Here are the results @mountiny.

### Mobile (iOS)

From testing on iOS it seems that the app performance is highly coupled to the amount of chats present. It is no surprise that it is influencing performance but the app should not slow down almost exponentially when amount of chats grows.

My first impression is that there are some heavy calculations based on entire data set going on in each item. I am going to investigate it further in some future, as it is out of the scope of the current RP.

Main branch:
Screenshot 2023-10-05 at 11 39 15
Current branch:
Screenshot 2023-10-05 at 11 38 40

### Web
For web, I was not able to load the app at all - no errors or anything particularly wrong. It after inspecting devtools it seems that there are a lot of calculations (probably related to what Matt and Monil discussed on Slack).

**Comment by mountiny** [comment]
Lets gooo

**Comment by mountiny** [comment]
@allroundexperts Thank you! Do you want to raise that here? https://expensify.slack.com/archives/C05LX9D6E07/p1696354046851689 it might be quicker to discuss and more people ot bound the ideas off

**Comment by kacper-mikolajczak** [comment]
Hi @allroundexperts, thanks for your feedback! 👍 
I agree with your concern about growing complexity and while I was thinking about the solution you've suggested, I came to the conclusion that it would eventually lead us to adding more abstraction that we might need at the moment.

In the current approach, we simply couple the OptionRowLHNData with the tiny wrapper in the form of OptionRowLHNDataWithFocus (naming itself is another topic - I am not sure it's correct). That's the whole abstraction we introduce.

In the suggested approach, we would first need to create withCurrentReportFocused HOC, but then we would still need to bound it to the OptionRowLHNData, which will lead us to creating OptionRowLHNDataWithFocus for this purpose. We just simply moved part of the logic (as shouldDisableFocusOptions still needs to be handled inside the wrapper component).

I agree that tightly coupling should be less preferred than composition, but the question is how much do we need to have withCurrentReportFocused abstraction as it is potentially just one line of extracted logic that seems kind of specific. Plus, it looks like it's going to be only used in one place and by sticking to YAGNI rule, we can think about leaving it as-is, wdyt?

Having that said, I am not as familiar with the codebase as you are, please let me know if I understood it right. Thanks a lot for the discussion :smile:

**Review by allroundexperts — APPROVED** [review]
As per the [Slack discussion](https://expensify.slack.com/archives/C05LX9D6E07/p1696354046851689), there was an agreement that we should continue using this pattern. As such, the changes are testing good!

**Review by mountiny — APPROVED** [review]
